### PR TITLE
MAINT: Translation of hyp2f1 for complex numbers into Cython, part 2

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

This PR does not close any issues known to me.
#### What does this implement/fix?

This PR is the third in an ongoing project to translate the Gaussian hypergeometric function (exposed through `scipy.special.hyp2f1`) into Cython. To recap: the original Fortran implementation has many defects but the difficulty of working with and reviewing many lines of dense and sparsely documented Fortran 77 code has thwarted previous efforts to correct them. The goal is to move parts of the implementation in Cython piece by piece while correcting the defects until at last a solid and maintainable implementation is completed.

This PR handles the following cases

1. Unital argument for c - a - b > 0 through [DLMF 15.4.20](https://dlmf.nist.gov/15.4#E20).
2. z = -1 and c = 1 + a - b through [DLMF 15.4.26](https://dlmf.nist.gov/15.4#E26).
3. c - a or c - b a negative integer by applying an Euler hypegeometric transformation [DLMF 15.8.1](https://dlmf.nist.gov/15.8#E1) after which the computation reduces to evaluation of a polynomial.
4. Re(z) >= 0 and |z| < 0.9 through the Maclaurin series for hyp2f1.

For cases 1 and 2, where the function can be evaluated as a ratio of products of gamma functions, the behavior has been updated to compute with logs and then exponentiate if an underflow or overflow occurs with the standard method.

We've also updated the `max_degree` argument of `hyp2f1_series` to be a uint64. This gets rid of platform dependent behavior for the cases where computation reduces to polynomial evaluation.

### CODE REVIEW

**Comment by steppi** [comment]
Interesting, one of the special cases is failing on 32 bit Linux.

```
scipy/special/tests/test_hyp2f1.py:298: in test_special_case_z_near_minus_1
    assert_allclose(hyp2f1(a, b, c, z), expected, rtol=rtol)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatch: 100%
E   Max absolute difference: 1.11460442e+27
E   Max relative difference: 2.4690097e+19
E    x: array(-1.114604e+27+0.j)
E    y: array(45143784.467839+0.j)
        a          = -102.1
        b          = -20.3
        c          = -80.8
        expected   = (45143784.46783885+0j)
        hyp2f1_test_case = Hyp2f1TestCase(a=-102.1, b=-20.3, c=-80.8, z=(-1+0j), expected=(45143784.46783885+0j), rtol=1e-07)
        rtol       = 1e-07
        self       = <scipy.special.tests.test_hyp2f1.TestHyp2f1 object at 0xd7f2c418>
        z          = (-1+0j)
```

The relevant code is
```python
        # Try again with logs if there has been an overflow.
        if zisnan(result) or zisinf(result) or result == 0.0:
            result = exp(
                lgam(c) - lgam(c - a) +
                lgam(c - a - b) - lgam(c - b)
            )
            result *= gammasgn(c) * gammasgn(c - a - b)
            result *= gammasgn(c - a) * gammasgn(c - b)
        return result
```

What's interesting is that the imaginary part of the incorrect computed answer is equal up to tolerance with the real part of the expected answer. I'll have to think about this. If I can't get it to work on 32 bit I think I should just scrap this idea to handle over/underflow by taking logs so that behavior can at least be consistent across platforms.

**Comment by steppi** [comment]
I was able to reproduce the error in a 32-bit Linux docker container. I might have time on Monday to try to get to the bottom of this.

**Comment by V0lantis** [comment]
Why not use the boost package to replace `cdflib` ? What do you think @rgommers

**Comment by steppi** [comment]
> Why not use the boost package to replace `cdflib` ? What do you think @rgommers

No particular reason. When I need low level access to a special function that's already in `scipy.special` I tend to default to whatever implementation is provided as a ufunc through `spipy/special/functions.json`. Note that replacing `Gamma` with the boost package will not fix the error. What's going on is that the correct branch isn't even being taken. DLMF 15.4.26 is applied when `zabs(z + 1) < 1e-15` and `fabs(1 + a - b - c) < 1e-15`. On 32 bit linux, `fabs(1 + a - b - c)` is evaluating to ~ 3.56e-15. 

One possible fix is just changing the tolerance for `fabs(1 + a - b - c)`, but I'd like to see that it doesn't hurt precision for other parameter values first.

**Comment by V0lantis** [comment]
> > Why not use the boost package to replace `cdflib` ? What do you think @rgommers
> 
> No particular reason. When I need low level access to a special function that's already in `scipy.special` I tend to default to whatever implementation is provided as a ufunc through `spipy/special/functions.json`. Note that replacing `Gamma` with the boost package will not fix the error. What's going on is that the correct branch isn't even being taken. DLMF 15.4.26 is applied when `zabs(z + 1) < 1e-15` and `fabs(1 + a - b - c) < 1e-15`. On 32 bit linux, `fabs(1 + a - b - c)` is evaluating to ~ 3.56e-15.
> 
> One possible fix is just changing the tolerance for `fabs(1 + a - b - c)`, but I'd like to see that it doesn't hurt precision for other parameter values first.

I am sorry, my question wasn't well asked. thanks to @mckib2, Scipy can now use boost C++ functions as in gh-13328. Since `hyp2f1` is currently not the only function in `cdflib` to not worry properly, I think it might be a better idea to replace the whole submodule by the boost implementation of those functions.

**Comment by steppi** [comment]
> I am sorry, my question wasn't well asked. thanks to @mckib2, Scipy can now use boost C++ functions as in [gh-13328](https://github.com/scipy/scipy/pull/13328). Since `hyp2f1` is currently not the only function in `cdflib` to not worry properly, I think it might be a better idea to replace the whole submodule by the boost implementation of those functions.

I see. hyp2f1 for complex numbers actually isn't implemented in cdflib, it's in specfun.f, so I thought you were referring to something else. I don't think hyp2f1 for complex z is implemented in boost actually.

**Comment by steppi** [comment]
@person142 I decided to punt on the error and mark that particular test as an xfail conditionally for 32 bit systems. That test doesn't even pass on any system with the original implementation so this is still an improvement.

File: scipy/special/_hyp2f1.pxd
Comment (person142) [inline scipy/special/_hyp2f1.pxd:123]: Note that long-term the right way to do this is to use the Lanczos approximation:

https://www.boost.org/doc/libs/1_38_0/libs/math/doc/sf_and_dist/html/math_toolkit/backgrounders/lanczos.html

to combine the large terms analytically. We've ported it to Cephes here:

https://github.com/scipy/scipy/blob/master/scipy/special/cephes/lanczos.c

File: scipy/special/_hyp2f1.pxd
Comment (person142) [inline scipy/special/_hyp2f1.pxd:138]: As above about Lanczos.

**Review by person142 — APPROVED** [review]
See comments about Lanczos approximation-it's ok if you don't want to do it now, but it's a very useful tool in any special functioner's toolbox. I can also go more in-depth on how to combine the terms if you like.

**Comment by mckib2** [comment]
> I see. hyp2f1 for complex numbers actually isn't implemented in cdflib, it's in specfun.f, so I thought you were referring to something else. I don't think hyp2f1 for complex z is implemented in boost actually.

As far as I'm aware, Boost distributions don't accept complex-valued inputs unfortunately

File: scipy/special/_hyp2f1.pxd
Comment (steppi) [reply scipy/special/_hyp2f1.pxd:123]: Nice! That looks really useful. I'll do this in my next PR.

**Comment by steppi** [comment]
@person142 I played around with using the Lanczos approximation on a different branch. It works well except for when any of `c`, `c - a`, or `c - b` are negative. I guess cases like that be can handled with the reflection formula? I haven't checked how the cancellation works out, but the case logic seems like it will be a little complicated. I'll think I'll hold off on this for a little bit and for the next few PRs start translating the different transforms and finish working on the fix for gh-8054 (which is almost done).

**Comment by steppi** [comment]
@person142 I noticed a bug. I made `max_degree` in `hyp2f1_series` a uint64, but the loop variable `k` is just an int.

**Comment by person142** [comment]
> It works well except for when any of c, c - a, or c - b are negative

Yes, the approximation doesn't work for negative real part; as you have observed you do need to use the reflection formula there.

**Comment by person142** [comment]
Thanks @steppi!

**Comment by steppi** [comment]
Before merging, I want to make `max_degree` signed after all. Since it's being mixed with doubles, the conversion from signed to double should be faster than unsigned to double in most cases.

**Comment by steppi** [comment]
Thanks @person142! I should be ready to submit a PR fixing gh-8054 next weekend. For the next PR, I think I may also change `max_degree` to a signed int64 instead of unsigned. From what I've read, I think the conversion from signed int to double may be faster on many processors, although I'm not sure if this almost 12 year old [stackoverflow answer](https://stackoverflow.com/questions/2043947/speed-difference-between-using-int-and-unsigned-int-when-mixed-with-doubles) is now out of date. I'll benchmark first though.

**Comment by person142** [comment]
Sounds good. For a complex function like `hpy2f1` I'd be surprised if the cast ended up dominating, but it never hurts to look.

**Pull request merged**
